File: examples/connext_dds/content_filtered_topic/java/cftSubscriber.java
/*******************************************************************************
 (c) 2005-2014 Copyright, Real-Time Innovations, Inc.  All rights reserved.
 RTI grants Licensee a license to use, modify, compile, and create derivative
 works of the Software.  Licensee has the right to distribute object form only
 for use with RTI products.  The Software is provided "as is", with no warranty
 of any type, including any warranty for fitness for any purpose. RTI is under
 no obligation to maintain or support the Software.  RTI shall not be liable for
 any incidental or consequential damages arising out of the use or inability to
 use the software.
 ******************************************************************************/

import java.util.Objects;

import com.rti.dds.domain.DomainParticipant;
import com.rti.dds.domain.DomainParticipantFactory;
import com.rti.dds.infrastructure.ConditionSeq;
import com.rti.dds.infrastructure.Duration_t;
import com.rti.dds.infrastructure.RETCODE_NO_DATA;
import com.rti.dds.infrastructure.RETCODE_TIMEOUT;
import com.rti.dds.infrastructure.ResourceLimitsQosPolicy;
import com.rti.dds.infrastructure.StatusKind;
import com.rti.dds.infrastructure.StringSeq;
import com.rti.dds.infrastructure.WaitSet;
import com.rti.dds.subscription.InstanceStateKind;
import com.rti.dds.subscription.ReadCondition;
import com.rti.dds.subscription.SampleInfo;
import com.rti.dds.subscription.SampleInfoSeq;
import com.rti.dds.subscription.SampleStateKind;
import com.rti.dds.subscription.Subscriber;
import com.rti.dds.subscription.ViewStateKind;
import com.rti.dds.topic.ContentFilteredTopic;
import com.rti.dds.topic.Topic;

// ===========================================================================

public class cftSubscriber extends Application implements AutoCloseable {
    private DomainParticipant participant = null;  // Usually one per
                                                   // application
    private cftDataReader reader = null;
    private final cftSeq dataSeq = new cftSeq();
    private final SampleInfoSeq infoSeq = new SampleInfoSeq();

    private int processData()
    {
        int samplesRead = 0;

        try {
            // Take available data from DataReader's queue
            reader.take(
                    dataSeq,
                    infoSeq,
                    ResourceLimitsQosPolicy.LENGTH_UNLIMITED,
                    SampleStateKind.ANY_SAMPLE_STATE,
                    ViewStateKind.ANY_VIEW_STATE,
                    InstanceStateKind.ANY_INSTANCE_STATE);

            for (int i = 0; i < dataSeq.size(); ++i) {
                SampleInfo info = infoSeq.get(i);

                if (info.valid_data) {
                    System.out.println("Received" + dataSeq.get(i));
                }
                samplesRead++;
            }
        } catch (RETCODE_NO_DATA noData) {
            // No data to process, not a problem
        } finally {
            // Data loaned from Connext for performance. Return loan when done.
            reader.return_loan(dataSeq, infoSeq);
        }

        return samplesRead;
    }

    private void runApplication()
    {
        // Start communicating in a domain
        participant = Objects.requireNonNull(
                DomainParticipantFactory.get_instance().create_participant(
                        getDomainId(),
                        DomainParticipantFactory.PARTICIPANT_QOS_DEFAULT,
                        null,  // listener
                        StatusKind.STATUS_MASK_NONE));

        // --- Create subscriber --- //
        /* To customize subscriber QoS, use
           the configuration file USER_QOS_PROFILES.xml */

        // A Subscriber allows an application to create one or more DataReaders
        Subscriber subscriber =
                Objects.requireNonNull(participant.create_subscriber(
                        DomainParticipant.SUBSCRIBER_QOS_DEFAULT,
                        null,  // listener
                        StatusKind.STATUS_MASK_NONE));

        // Register the datatype to use when creating the Topic
        String typeName = cftTypeSupport.get_type_name();
        cftTypeSupport.register_type(participant, typeName);

        // Create a Topic with a name and a datatype
        Topic topic = Objects.requireNonNull(participant.create_topic(
                "Example cft",
                typeName,
                DomainParticipant.TOPIC_QOS_DEFAULT,
                null,  // listener
                StatusKind.STATUS_MASK_NONE));

        // --- Create reader --- //

        // The default parameter list that we will include in the
        // sequence of parameters will be "1","4" (i.e., 1 <= x <= 4).
        String param_list[] = { "1", "4" };
        // Sequence of parameters for the content filter expression */
        StringSeq parameters =
                new StringSeq(java.util.Arrays.asList(param_list));

        /* Create the content filtered topic in case isCftSelected
         * is true.
         * The Content Filter Expresion has two parameters:
         * - %0 -- x must be greater or equal than %0.
         * - %1 -- x must be less or equal than %1.
         */
        ContentFilteredTopic cft = null;
        if (isCftSelected()) {
            cft = participant.create_contentfilteredtopic(
                    "ContentFilteredTopic",
                    topic,
                    "(x >= %0 and x <= %1)",
                    parameters);

            System.out.print("Using ContentFiltered Topic\n");
            reader = (cftDataReader) Objects.requireNonNull(
                    subscriber.create_datareader(
                            cft,
                            Subscriber.DATAREADER_QOS_DEFAULT,
                            null,
                            StatusKind.STATUS_MASK_ALL));
            System.out.print("\n==========================\n");
            System.out.print("Using CFT\nFilter: 1 <= x <= 4\n");
            System.out.print("===========================\n");
        } else {
            System.out.print("Using Normal Topic\n");
            reader = (cftDataReader) Objects.requireNonNull(
                    subscriber.create_datareader(
                            topic,
                            Subscriber.DATAREADER_QOS_DEFAULT,
                            null,
                            StatusKind.STATUS_MASK_ALL));
        }

        /* If you want to set the reliability and history QoS settings
         * programmatically rather than using the XML, you will need to add
         * the following lines to your code and comment out the
         * create_datareader calls above.
         */

        /*
        DataReaderQos datareader_qos = new DataReaderQos();
        subscriber.get_default_datareader_qos(datareader_qos);

        datareader_qos.reliability.kind =
            ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS;
        datareader_qos.durability.kind =
            DurabilityQosPolicyKind.TRANSIENT_LOCAL_DURABILITY_QOS;
        datareader_qos.history.kind =
            HistoryQosPolicyKind.KEEP_LAST_HISTORY_QOS;
        datareader_qos.history.depth = 20;

        if (isCftSelected()) {
            System.out.print("Using ContentFiltered Topic\n");
            reader = (cftDataReader) Objects.requireNonNull(
                subscriber.create_datareader(
                    cft, datareader_qos, listener,
                    StatusKind.STATUS_MASK_ALL));
            System.out.print("\n==========================\n");
            System.out.print("Using CFT\nFilter: 1 <= x <= 4\n");
            System.out.print("===========================\n");

        } else {
            System.out.print("Using Normal Topic\n");
            reader = (cftDataReader) Objects.requireNonNull(
                subscriber.create_datareader(
                    topic, datareader_qos, listener,
                    StatusKind.STATUS_MASK_ALL));
        }
        */

        // --- Wait for data --- //

        // Create ReadCondition that triggers when data in reader's queue
        ReadCondition condition = reader.create_readcondition(
                SampleStateKind.ANY_SAMPLE_STATE,
                ViewStateKind.ANY_VIEW_STATE,
                InstanceStateKind.ANY_INSTANCE_STATE);

        // WaitSet will be woken when the attached condition is triggered, or
        // timeout
        WaitSet waitset = new WaitSet();
        waitset.attach_condition(condition);
        final Duration_t waitTimeout = new Duration_t(1, 0);

        int samplesRead = 0;
        int iterationCount = 0;
        ConditionSeq activeConditions = new ConditionSeq();

        // Main loop. Wait for data to arrive and process when it arrives
        while (!isShutdownRequested() && samplesRead < getMaxSampleCount()) {
            /* If we are not using CFT, we do not need to change the filter
             * parameters */
            if (iterationCount == 10 && isCftSelected()) {
                System.out.print("\n==========================\n");
                System.out.print("Changing filter parameters\n");
                System.out.print("Filter: 5 <= x <= 9\n");
                System.out.print("===========================\n");

                parameters.set(0, "5");
                parameters.set(1, "9");

                cft.set_expression_parameters(parameters);

            } else if (iterationCount == 20 && isCftSelected()) {
                System.out.print("\n==========================\n");
                System.out.print("Changing filter parameters\n");
                System.out.print("Filter: 3 <= x <= 9\n");
                System.out.print("===========================\n");

                StringSeq oldParameters = new StringSeq();
                cft.get_expression_parameters(oldParameters);

                oldParameters.set(0, "3");
                cft.set_expression_parameters(oldParameters);
            }

            try {
                // Wait fills in activeConditions or times out
                waitset.wait(activeConditions, waitTimeout);

                // Read condition triggered, process data
                samplesRead += processData();

            } catch (RETCODE_TIMEOUT timeout) {
            }

            iterationCount++;
        }
    }

    @Override public void close()
    {
        // Delete all entities (DataReader, Topic, Subscriber,
        // DomainParticipant)
        if (participant != null) {
            participant.delete_contained_entities();

            DomainParticipantFactory.get_instance().delete_participant(
                    participant);
        }
    }

    public static void main(String[] args)
    {
        // Create example and run: Uses try-with-resources,
        // subscriberApplication.close() automatically called
        try (cftSubscriber subscriberApplication = new cftSubscriber()) {
            subscriberApplication.parseArguments(
                    args,
                    ApplicationType.SUBSCRIBER);
            subscriberApplication.addShutdownHook();
            subscriberApplication.runApplication();
        }

        // Releases the memory used by the participant factory. Optional at
        // application exit.
        DomainParticipantFactory.finalize_instance();
    }
}

File: examples/connext_dds/content_filtered_topic_string_filter/c++98/cft_subscriber.cxx
/*******************************************************************************
 (c) 2005-2014 Copyright, Real-Time Innovations, Inc.  All rights reserved.
 RTI grants Licensee a license to use, modify, compile, and create derivative
 works of the Software.  Licensee has the right to distribute object form only
 for use with RTI products.  The Software is provided "as is", with no warranty
 of any type, including any warranty for fitness for any purpose. RTI is under
 no obligation to maintain or support the Software.  RTI shall not be liable for
 any incidental or consequential damages arising out of the use or inability to
 use the software.
 ******************************************************************************/

#include <iostream>
#include <stdio.h>
#include <stdlib.h>

#include "cft.h"
#include "cftSupport.h"
#include "ndds/ndds_cpp.h"
#include "application.h"

using namespace application;

static int shutdown_participant(
        DDSDomainParticipant *participant,
        const char *shutdown_message,
        int status);

unsigned int process_data(cftDataReader *typed_reader)
{
    cftSeq data_seq;
    DDS_SampleInfoSeq info_seq;
    unsigned int samples_read = 0;

    typed_reader->take(
            data_seq,
            info_seq,
            DDS_LENGTH_UNLIMITED,
            DDS_ANY_SAMPLE_STATE,
            DDS_ANY_VIEW_STATE,
            DDS_ANY_INSTANCE_STATE);

    for (int i = 0; i < data_seq.length(); ++i) {
        if (info_seq[i].valid_data) {
            cftTypeSupport::print_data(&data_seq[i]);
            samples_read++;
        }
    }

    // Data loaned from Connext for performance. Return loan when done.
    DDS_ReturnCode_t retcode = typed_reader->return_loan(data_seq, info_seq);
    if (retcode != DDS_RETCODE_OK) {
        std::cerr << "return loan error " << retcode << std::endl;
    }

    return samples_read;
}

int run_subscriber_application(
        unsigned int domain_id,
        unsigned int sample_count,
        bool normal_topic)
{
    DDS_Duration_t wait_timeout = { 1, 0 };

    /* For this filter we only allow 1 parameter*/
    DDS_StringSeq parameters(1);
    const char *param_list[] = { "SOME_STRING" };

    // Start communicating in a domain, usually one participant per application
    DDSDomainParticipant *participant =
            DDSTheParticipantFactory->create_participant(
                    domain_id,
                    DDS_PARTICIPANT_QOS_DEFAULT,
                    NULL /* listener */,
                    DDS_STATUS_MASK_NONE);
    if (participant == NULL) {
        return shutdown_participant(
                participant,
                "create_participant error",
                EXIT_FAILURE);
    }

    // A Subscriber allows an application to create one or more DataReaders
    DDSSubscriber *subscriber = participant->create_subscriber(
            DDS_SUBSCRIBER_QOS_DEFAULT,
            NULL /* listener */,
            DDS_STATUS_MASK_NONE);
    if (subscriber == NULL) {
        return shutdown_participant(
                participant,
                "create_subscriber error",
                EXIT_FAILURE);
    }

    // Register the datatype to use when creating the Topic
    const char *type_name = cftTypeSupport::get_type_name();
    DDS_ReturnCode_t retcode =
            cftTypeSupport::register_type(participant, type_name);
    if (retcode != DDS_RETCODE_OK) {
        return shutdown_participant(
                participant,
                "register_type error",
                EXIT_FAILURE);
    }

    // Create a Topic with a name and a datatype
    DDSTopic *topic = participant->create_topic(
            "Example cft",
            type_name,
            DDS_TOPIC_QOS_DEFAULT,
            NULL /* listener */,
            DDS_STATUS_MASK_NONE);
    if (topic == NULL) {
        return shutdown_participant(
                participant,
                "create_topic error",
                EXIT_FAILURE);
    }

    parameters.from_array(param_list, 1);

    DDSContentFilteredTopic *cft = NULL;
    if (!normal_topic) {
        /* create_contentfilteredtopic_with_filter */
        cft = participant->create_contentfilteredtopic_with_filter(
                "ContentFilteredTopic",
                topic,
                "name MATCH %0",
                parameters,
                DDS_STRINGMATCHFILTER_NAME);
        if (cft == NULL) {
            return shutdown_participant(
                    participant,
                    "create_contentfilteredtopic error",
                    EXIT_FAILURE);
        }
    }

    /* Here we create the reader either using a Content Filtered Topic or
     * a normal topic */
    DDSDataReader *untyped_reader = NULL;
    if (!normal_topic) {
        std::cout << "Using ContentFiltered Topic\n";
        untyped_reader = subscriber->create_datareader(
                cft,
                DDS_DATAREADER_QOS_DEFAULT,
                NULL,
                DDS_STATUS_MASK_ALL);
    } else {
        std::cout << "Using Normal Topic\n";
        untyped_reader = subscriber->create_datareader(
                topic,
                DDS_DATAREADER_QOS_DEFAULT,
                NULL,
                DDS_STATUS_MASK_ALL);
    }
    if (untyped_reader == NULL) {
        return shutdown_participant(
                participant,
                "create_datareader error",
                EXIT_FAILURE);
    }


    /* If you want to set the reliability and history QoS settings
     * programmatically rather than using the XML, you will need to add
     * the following lines to your code and comment out the
     * create_datareader calls above.
     */
    /*
        DDS_DataReaderQos datareader_qos;
        retcode = subscriber->get_default_datareader_qos(datareader_qos);
        if (retcode != DDS_RETCODE_OK) {
            return shutdown_participant(
                    participant,
                    "get_default_datareader_qos error",
                    EXIT_FAILURE);
        }

        datareader_qos.reliability.kind = DDS_RELIABLE_RELIABILITY_QOS;
        datareader_qos.durability.kind = DDS_TRANSIENT_LOCAL_DURABILITY_QOS;
        datareader_qos.history.kind = DDS_KEEP_LAST_HISTORY_QOS;
        datareader_qos.history.depth = 20;

        if (!normal_topic) {
            std::cout << "Using ContentFiltered Topic\n";
            untyped_reader = subscriber->create_datareader(
                cft, datareader_qos, NULL,
                DDS_STATUS_MASK_ALL);
        } else {
            std::cout << "Using Normal Topic\n";
            untyped_reader = subscriber->create_datareader(
                topic, datareader_qos, NULL,
                DDS_STATUS_MASK_ALL);
        }
        if (untyped_reader == NULL) {
            return shutdown_participant(
                    participant,
                    "create_datareader error",
                    EXIT_FAILURE);
        }

    */
    /* Change the filter */
    if (!normal_topic) {
        printf(">>> Now setting a new filter: name MATCH \"EVEN\"\n");
        retcode = cft->append_to_expression_parameter(0, "EVEN");
        if (retcode != DDS_RETCODE_OK) {
            return shutdown_participant(
                    participant,
                    "append_to_expression_parameter error",
                    EXIT_FAILURE);
        }
    }

    // Narrow casts from a untyped DataReader to a reader of your type
    cftDataReader *typed_reader = cftDataReader::narrow(untyped_reader);
    if (typed_reader == NULL) {
        return shutdown_participant(
                participant,
                "DataReader narrow error",
                EXIT_FAILURE);
    }

    // Create ReadCondition that triggers when unread data in reader's queue
    DDSReadCondition *read_condition = typed_reader->create_readcondition(
            DDS_NOT_READ_SAMPLE_STATE,
            DDS_ANY_VIEW_STATE,
            DDS_ANY_INSTANCE_STATE);
    if (read_condition == NULL) {
        return shutdown_participant(
                participant,
                "create_readcondition error",
                EXIT_FAILURE);
    }

    // WaitSet will be woken when the attached condition is triggered
    DDSWaitSet waitset;
    retcode = waitset.attach_condition(read_condition);
    if (retcode != DDS_RETCODE_OK) {
        return shutdown_participant(
                participant,
                "attach_condition error",
                EXIT_FAILURE);
    }

    // Main loop. Wait for data to arrive, and process when it arrives
    unsigned int samples_read = 0;
    while (!shutdown_requested && samples_read < sample_count) {
        DDSConditionSeq active_conditions_seq;

        // Wait for data
        retcode = waitset.wait(active_conditions_seq, wait_timeout);

        if (retcode == DDS_RETCODE_OK) {
            // If the read condition is triggered, process data
            samples_read += process_data(typed_reader);
        }

        if (normal_topic) {
            continue;
        }
        if (samples_read == 10) {
            std::cout << "\n===========================\n";
            std::cout << "Changing filter parameters\n";
            std::cout << "Append 'ODD' filter\n";
            std::cout << "===========================\n";
            retcode = cft->append_to_expression_parameter(0, "ODD");
            if (retcode != DDS_RETCODE_OK) {
                return shutdown_participant(
                        participant,
                        "append_to_expression_parameter error",
                        EXIT_FAILURE);
            }
        } else if (samples_read == 20) {
            std::cout << "\n===========================\n";
            std::cout << "Changing filter parameters\n";
            std::cout << "Removing 'EVEN' filter \n";
            std::cout << "===========================\n";

            retcode = cft->remove_from_expression_parameter(0, "EVEN");
            if (retcode != DDS_RETCODE_OK) {
                return shutdown_participant(
                        participant,
                        "append_to_expression_parameter error",
                        EXIT_FAILURE);
            }
        }
    }

    // Cleanup
    return shutdown_participant(participant, "Shutting down", 0);
}

// Delete all entities
static int shutdown_participant(
        DDSDomainParticipant *participant,
        const char *shutdown_message,
        int status)
{
    DDS_ReturnCode_t retcode;

    std::cout << shutdown_message << std::endl;

    if (participant != NULL) {
        // Cleanup everything created by this Participant
        retcode = participant->delete_contained_entities();
        if (retcode != DDS_RETCODE_OK) {
            std::cerr << "delete_contained_entities error" << retcode
                      << std::endl;
            status = EXIT_FAILURE;
        }

        retcode = DDSTheParticipantFactory->delete_participant(participant);
        if (retcode != DDS_RETCODE_OK) {
            std::cerr << "delete_participant error" << retcode << std::endl;
            status = EXIT_FAILURE;
        }
    }
    return status;
}

int main(int argc, char *argv[])
{
    // Parse arguments and handle control-C
    ApplicationArguments arguments;
    parse_arguments(arguments, argc, argv, subscriber);
    if (arguments.parse_result == PARSE_RETURN_EXIT) {
        return EXIT_SUCCESS;
    } else if (arguments.parse_result == PARSE_RETURN_FAILURE) {
        return EXIT_FAILURE;
    }
    setup_signal_handlers();

    // Sets Connext verbosity to help debugging
    NDDSConfigLogger::get_instance()->set_verbosity(arguments.verbosity);

    int status = run_subscriber_application(
            arguments.domain_id,
            arguments.sample_count,
            arguments.normal_topic);

    // Releases the memory used by the participant factory.  Optional at
    // application exit
    DDS_ReturnCode_t retcode = DDSDomainParticipantFactory::finalize_instance();
    if (retcode != DDS_RETCODE_OK) {
        std::cerr << "finalize_instance error" << retcode << std::endl;
        status = EXIT_FAILURE;
    }

    return status;
}
